File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int max(int a,int b){
        if(a>=b) return a;
        else return b;
    }
    public int min(int a,int b ){
        if(a<=b) return a;
        else return b;
    }
    public int trap(int[] height) {
        int[] prefix= new int[height.length];
        int[] suffix=new int[height.length];
        int res=0;
        prefix[0]=height[0];
        suffix[height.length-1]=height[height.length-1];
        for(int i=1;i<height.length;i++){
            prefix[i]=max(prefix[i-1],height[i]);
        }
        for(int i=height.length-2;i>=0;i--){
            suffix[i]=max(suffix[i+1],height[i]);
        }
        for(int i=0;i<height.length;i++){
            res+=min(prefix[i],suffix[i])-height[i];
        }
        return res;

        
    }
}

File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int i = 0;
        int j = height.size()-1;
        int leftMax = 0;
        int rightMax= 0;
        int total =0;
        while(i<j){
            leftMax = max(leftMax,height[i]);
            rightMax = max(rightMax,height[j]);
            if(leftMax<=rightMax){
                total+=leftMax-height[i++];
            }
            else{
                total+=rightMax-height[j--];
            }
        }
        return total;

        
    }
};